### ISSUE TITLE
Keeping headerbar/sidemenu in place

### ISSUE THREAD

**Issue by surma** [jakearchibald/navigation-transitions#9]
I was wondering if and how this API could implement a static (and even functional!) headerbar/sidemenu throughout transitions like shown in [this video (450KB)](https://f.surma.link/tmp/headerbar_web.mp4). I’d like this API to be able to do this, but not sure if there’s any implications I am missing.

Taking some inspiration from the [“Immediate Slide in/out transition” example](https://github.com/jakearchibald/navigation-transitions/blob/fc28acb003c3b65515df3690b81962ac100c61f1/README.md#immediate-slide-inout-transition), this is what I came up with:

``` js
window.addEventListener('navigate', event => {
  if (event.type == 'reload') return;

  const newURL = new URL(event.url);
  if (newURL.origin != location.origin) return;

  const oldContent = document.querySelector('#content');
  // Does some kind of animation so that the content div is not visible anymore
  const outAnimation = animateOut(oldContent);

  event.transitionUntil(async function() {
    await outAnimation;
    const newWin = await event.newWindow;
    const newContent = newWin.document.documentElement.querySelector('#content');
    // Styles that move content out-of-view so it can be animated in
    newContent.classList.add('about-to-be-animated-in');
    oldContent.parentNode.replaceChild(newContent, oldContent);
    return animateIn(newContent);
  }());
});
```

Would this work? Should this work?

@jakearchibald @paullewis WDYT?

**Comment by jakearchibald** [comment]
I think that would work, as long as you:
- waited for `#content` to be in the new document's DOM, eg wait for `document.interactive`
- deep-cloned `newContent` before appending it, else it's taken out of the new document
- cater for the scroll position of the new document being different to the current document

It might be simpler to `z-index` the old document to the top, make it's `documentElement` transparent, then slide-to-the-left the _content_ area of the old doc, while sliding-from-the-right the new `documentElement`.

This way you're not copying DOM, you don't have to worry about load state of the new doc, and don't have to worry about scroll position.

**Comment by surma** [comment]
> deep-cloned newContent before appending it, else it's taken out of the new document

That was actually intentional. See below.

> It might be simpler to z-index the old document to the top, make it's documentElement transparent, then slide-to-the-left the content area of the old doc, while sliding-from-the-right the new documentElement.

Imagine the user interacting with a sidebar while waiting for the new content. We don’t want to just rip out the open sidebar and fade into the closed version of the side bar.

That’s why I was wondering if I could adopt individual nodes from the new doc to the old one.

**Comment by jakearchibald** [comment]
Ahh, so you'd move them back at the end of the transition?

Something that's been bouncing around my head for a while that I want to know the answer to is: What happens when you take a node from one document to another while that node's being populated by the html parser?

Moving the node around is risky if the new document does `document.querySelector('.side-bar')` and it's not there, but I guess you'd just have to defend against that.

**Comment by surma** [comment]
> Ahh, so you'd move them back at the end of the transition?

Not quite sure what you mean. I am effectively discarding parts of the `newWin.document`, only ripping out the parts I need (i.e. `#content`) and moving it to the old `document`.

> Something that's been bouncing around my head for a while that I want to know the answer to is: What happens when you take a node from one document to another while that node's being populated by the html parser?

_If_ the browser keeps populating that note by reference, regardless which document it is attached to, you could do very interesting stuff like having the sidebar persisted and stream huge documents into the document. If that is currently not specified, might be worth talking to browser people/WHATWG about it. (Maybe create a separate issue?)

> Moving the node around is risky if the new document does document.querySelector('.side-bar') and it's not there, but I guess you'd just have to defend against that.

Yes, I’d say that is something that needs to be accounted for in the app’s architecture and the developer’s code.

**Comment by jakearchibald** [comment]
> Not quite sure what you mean. I am effectively discarding parts of the `newWin.document`, only ripping out the parts I need (i.e. `#content`) and moving it to the old `document`.

Oh, that isn't possible in the current model. Once `transitionUntil(p)` is resolved the transition is complete, the new document becomes the active document and the old one is destroyed (or put into the bf cache depending on the browser).

**Comment by surma** [comment]
Than my question becomes: “Should this be possible?” 

Another approach would be to migrate the sidebar from the old `document` to the new one, to keep it’s state and everything in tact.

**Comment by jakearchibald** [comment]
I think we need to look at the use case, but if we're not going to move from the current document to a new one you're circumventing so much of regular browser navigation you may as well use `pushState`.

**Comment by jakearchibald** [comment]
I think you want `<iframe seamless>` 😢

**Comment by surma** [comment]
That’s fair. If persisting elements across navigations is where we draw the line of responsibility, I can live with that. Seems like a easy enough mantra.

Closing this for now.

**Comment by jakearchibald** [comment]
Let's keep this open because the overall transition is something we should have an example for. As in, persistent header that retains interactivity.

If it isn't possible we need to have a think about whether it should be.

**Comment by joegaffey** [comment]
I feel this would be necessary for my use case, a web app suite with an interactive cross-app header bar.
